block/src/Pos/Block/Logic/Header.hs

```diff
@@ -352,29 +357,39 @@ getHeadersOlderExp upto = do
                 | otherwise = selGo es ii $ succ skipped
         in selGo elems ixs 0
 
+
+data GetHashesRangeException = GetHashesRangeException Text deriving (Show,Generic)
+instance Exception GetHashesRangeException
+
+-- | Throws 'GetHashesRangeException'.
+throwGHR :: MonadThrow m => Text -> m a
+throwGHR = throwM . GetHashesRangeException
+
 -- | Given optional @depthLimit@, @from@ and @to@ headers where @from@
 -- is older (not strict) than @to@, and valid chain in between can be
 -- found, headers in range @[from..to]@ will be found. If the number
 -- of headers in the chain (which should be returned) is more than
 -- @depthLimit@, error will be thrown.
-getHeadersRange ::
+getHashesRange ::
        forall m. (HasConfiguration, MonadDBRead m)
     => Maybe Word
     -> HeaderHash
     -> HeaderHash
-    -> m (Either Text (OldestFirst NE HeaderHash))
-getHeadersRange depthLimitM older newer | older == newer = runExceptT $ do
+    -> m (OldestFirst NE HeaderHash)
+getHashesRange depthLimitM older newer | older == newer = do
     unlessM (isJust <$> DB.getHeader newer) $
-        throwError "getHeadersRange: can't find newer-older header"
+        throwGHR "can't find newer-older header"
```

Comment:
Is this really an error? Isn't it a completely normal thing that can be handled in a non-exceptional way?

Reply:
The updated code follows the guidelines in https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md unlike the old code.

The reasoning behind the current design choice in the guidelines is that we want only a single way to throw exceptions in any part of code - it could be `ExceptT` in pure code, or `MonadThrow` in (potentially) impure code. This guarantees that we can safely use `bracket` everywhere, and that there's never need to use both `catch` and `catchError`.

This code was using `throwError`, but it wasn't pure, so it was rightly rewritten. If you disagree with the use of `throwM` here, the alternative is to return `Either`, but without `runExceptT` which creates a secondary exit path from any part of the code.

Additionally, we don't use `Text` to represent errors (or "normal things", as you refer to them above) anymore, and the updated code at least uses a newtype (ideally it should be an ADT).

Reply:
> Additionally, we don't use Text to represent errors (or "normal things", as you refer to them above) anymore, and the updated code at least uses a newtype (ideally it should be an ADT).

Sure I'm all for that. In principle, though, I think many of the `throw*`s seen in this patch aren't really exceptions, they're normal outcomes. This line in particular: a header hash was given, but it's not in the database. Ok, that's an expected outcome, and the caller can deal with it normally. We wouldn't promise the caller a total function from header hash to blocks, so we should't pretend we have by giving such a type.

Reply:
The normal outcomes that short-circuit the computation (thus can be, theoretically, represented by `MonadThrow` or `MonadError`) go under the name of "regular errors" in the guidelines. You might argue that it's a misnomer, but it doesn't change the essence of the question.

Reply:
It's confusing. Here I have some `m` that could be and probably is `IO`-capable. What does "throwing" mean exactly? It suggests an exception, as in a real true exception (the kind that only makes sense in `IO`) but it could also be a "normal" short-circuit.

Anyway, it's beyond the scope of this patch I guess. I think the true issue is the lack of good separation of pure and impure programs. If we had that, then there would be no ambiguity about what "throwing" means.

Reply:
@avieth this patch tries to fix problems that were missed in PR 2154: https://github.com/input-output-hk/cardano-sl/pull/2154/#discussion_r163384764

* Code is violating exceptions guideline
* It wasn't fixed, so now it is

My patch may be confusing, but it is only because exceptions handling is hard. In this particular place there is no good separation between request exceptions and everything else, pure vs impure, so I just follow the guideline written by @int-index. You're more than welcomed to read and argue with it, but for now -- standards first.

Reply:
Right I won't let it stand in the way of merging this, it's a separate issue. It's not really about exception handling guidelines, it's about when to use exceptions at all.

Reply:
I take it back, this *is* a problem. Discussion here https://github.com/input-output-hk/cardano-sl/pull/2492